### PR TITLE
HADOOP-13980. fsck - Work In Progress

### PR DESCRIPTION
**Pull request by bgaborg**
WIP

### CODE REVIEW

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsckViolationHandler.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+import org.apache.commons.math3.ode.UnknownParameterException;
+import org.apache.hadoop.fs.s3a.S3AFileStatus;
+import org.apache.hadoop.fs.s3a.S3AFileSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Violation handler for the S3Guard's fsck
+ *
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsckViolationHandler.java]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 31 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1056 | trunk passed |
| +1 | compile | 33 | trunk passed |
| +1 | checkstyle | 20 | trunk passed |
| +1 | mvnsite | 34 | trunk passed |
| +1 | shadedclient | 695 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | trunk passed |
| 0 | spotbugs | 57 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 56 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 30 | the patch passed |
| +1 | javac | 30 | the patch passed |
| -0 | checkstyle | 6 | The patch fails to run checkstyle in hadoop-aws |
| -1 | mvnsite | 18 | hadoop-aws in the patch failed. |
| -1 | whitespace | 0 | The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | shadedclient | 736 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | the patch passed |
| -1 | findbugs | 62 | hadoop-tools/hadoop-aws generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
||| _ Other Tests _ |
| +1 | unit | 282 | hadoop-aws in the patch passed. |
| -1 | asflicense | 28 | The patch generated 1 ASF License warnings. |
| | | 3249 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-tools/hadoop-aws |
|  |  Comparison of String objects using == or != in org.apache.hadoop.fs.s3a.s3guard.S3GuardFsck.compareFileStatusToPathMetadata(S3AFileStatus, PathMetadata)   At S3GuardFsck.java:== or != in org.apache.hadoop.fs.s3a.s3guard.S3GuardFsck.compareFileStatusToPathMetadata(S3AFileStatus, PathMetadata)   At S3GuardFsck.java:[line 221] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1060 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 27a7f95ca05a 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ec851e4 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/artifact/out//home/jenkins/jenkins-slave/workspace/hadoop-multibranch_PR-1060/out/maven-patch-checkstyle-hadoop-tools_hadoop-aws.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/artifact/out/patch-mvnsite-hadoop-tools_hadoop-aws.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/artifact/out/whitespace-eol.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/artifact/out/new-findbugs-hadoop-tools_hadoop-aws.html |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 447 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1060/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTableAccess.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+  private final Table table;
+
+  S3GuardTableAccess(final DynamoDBMetadataStore store) {
+    this.store = checkNotNull(store);
+    this.table = store.getTable();
+  }
+
+  private String getUsername() {
+    return store.getUsername();
```
Comment (mackrorysd) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTableAccess.java]: Depending on DynamoDB-specifics here. Mildly dangerous.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsck.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    final Path path = s3FileStatus.getPath();
+    System.out.println("== Path: " + path);
+    ComparePair comparePair = new ComparePair(s3FileStatus, msPathMetadata);
+
+    if (!path.equals(path(ROOT_PATH_STRING))) {
+      final Path parentPath = path.getParent();
+      final PathMetadata parentPm = metadataStore.get(parentPath);
+
+      if (parentPm == null) {
+        comparePair.violations.add(Violation.NO_PARENT_ENTRY);
```
Comment (mackrorysd) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsck.java]: I think this approach is okay. I had pictured it being more pluggable and modular, but I would imagine that would be woefully inefficient. I think we just need to make sure this code stays as clean & readable as it is, and that will get harder over time as we add new problems we come across.

**Pull request closed without merging**